text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+readability, and allows reordering instructions without changing the argument order.
+
+We can further refine the above example to avoid the `mov` instruction:
+
+```rust
+let mut x: u32 = 3;
+unsafe {
+    asm!("add {0}, {number}", inout(reg) x, number = imm 5);
+}
+assert_eq!(x, 8);
+```
+
+We can see that `inout` is used to specify an argument that is both input and output.
+This is different from specifying an input and output separately in that it is guaranteed to assign both to the same register.
+
+It is also possible to specify different variables for the input and output parts of an `inout` operand:
+
+```rust
+let x: u32 = 3;
+let y: u32;
+unsafe {
+    asm!("add {0}, {number}", inout(reg) x => y, number = imm 5);
+}
+assert_eq!(y, 8);
+```
+
+## Late output operands
+
+The Rust compiler is conservative with its allocation of operands. It is assumed that an `out`
+can be written at any time, and can therefore not share its location with any other argument.
+However, to guarantee optimal performance it is important to use as few registers as possible,
+so they won't have to be saved and reloaded around the inline assembly block.
+To achieve this Rust provides a `lateout` specifier. This can be used on any output that is
+guaranteed to be written only after all inputs have been consumed.
+There is also a `inlateout` variant of this specifier.
+
+Here is an example where `inlateout` *cannot* be used:
+
+```rust
+let mut a = 4;
+let b = 4;
+let c = 4;
+unsafe {
+    asm!("
+        add {0}, {1}
+        add {0}, {2}
+    ", inout(reg) a, in(reg) b, in(reg) c);
+}
+assert_eq!(a, 12);
+```
+
+Here the compiler is free to allocate the same register for inputs `b` and `c` since it knows they have the same value. However it must allocate a separate register for `a` since it uses `inout` and not `inlateout`.
+
+However the following example can use `inlateout` since the output is only modified after all input registers have been read:
+
+```rust
+let mut a = 4;
+let b = 4;
+unsafe {
+    asm!("add {0}, {1}", inlateout(reg) a, in(reg) b);
+}
+assert_eq!(a, 8);
+```
+
+As you can see, this assembly fragment will still work correctly if `a` and `b` are assigned to the same register.
+
+## Explicit register operands
+
+Some instructions require that the operands be in a specific register.
+Therefore, Rust inline assembly provides some more specific constraint specifiers.
+While `reg` is generally available on any architecture, these are highly architecture specific. E.g. for x86 the general purpose registers `eax`, `ebx`, `ecx`, `edx`, `ebp`, `esi`, and `edi`
+among others can be addressed by their name.
+
+```rust
+unsafe {
+    asm!("out 0x64, {}", in("eax") cmd);
+}
+```
+
+In this example we call the `out` instruction to output the content of the `cmd` variable
+to port `0x64`. Since the `out` instruction only accepts `eax` (and its sub registers) as operand
+we had to use the `eax` constraint specifier.
```

Comment:
One of the problems I have with the current `asm!()` macro is that it does not at all feel like the error reporting story is on the same level of quality as the rest of `rustc`. See e.g. rust-lang/rust#15402

If I accidentally use `reg` here instead of `"eax"`, will I get poor error messages from deep inside the belly of LLVM?

Reply:
Worth noting that your specific example won't be possible anymore, since as noted in the RFC, the input/output parameters will only accept raw numeric types (and pointers where it makes sense). Your wrapper struct would get rejected by the type checker.

There are other types of weird "low-level" llvm errors that can get reported when the feature is misused though. Notably, I remember getting really weird aborts when misusing constraints. Rustc will probably need to sanitize them before passing it to LLVM, making sure they make sense. Some things on the top of my head:

- Using the same named register twice as an input parameter (it's UB, LLVM accepts it and will either abort or does something weird)
- Using the same named register as both a clobber and an input or an output (UB, LLVM generates garbage, sometimes aborts)

Reply:
The specification in this RFC and the validation it requires should eliminate the possibility of any LLVM internal error. For example, the RFC explicitly says this:

> It is a compile-time error to use the same explicit register two input operand or two output operands.

Reply:
In this case, the `out` instruction must use the `al`, `ax` or `eax` register per the instruction set reference.

If I accidentally wrote `reg` instead of `"eax"` in the example, would the error message be of the same quality as normal rust error messages (or at least using the same infrastructure wrt highlighting the span in question etc)?

Another concern is that it would potentially compile some of the time, when LLVM happens to allocate `eax` as its register of choice. Would that be possible?

Reply:
> If I accidentally wrote `reg` instead of `"eax"` in the example, would the error message be of the same quality as normal rust error messages (or at least using the same infrastructure wrt highlighting the span in question etc)?

LLVM has support for sending error messages from inline assembly back to rustc so that they can be displayed through rustc's normal error message functionality. This already works with the current `asm!` macro: https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e5d9c3c74edc6e02858ec965abfd4d98

> Another concern is that it would potentially compile some of the time, when LLVM happens to allocate `eax` as its register of choice. Would that be possible?

Yes it would sometimes compile fine and sometimes not, depending on the register selected by LLVM. There isn't much we can do about that.

Reply:
More than cryptic error messages I fear inefficiency and silent errors without any error messages.

First issue...

Here's one example: https://godbolt.org/z/sT_8Qv

short foo(short input) {
    short result;
    asm("xchg %h0, %b0" : "=Q"(result) : "0"(input));
    return result;
}

How this simple function could be expressed in new inline assembler form? I couldn't see a way to do that.

Second issue is more subtle: ttps://godbolt.org/z/5d0MxL

As you can see if you would pass one-byte variable and request an "r" constraint (instead of "q" or "Q" constraint) LLVM is all to happy to make a mess out of your assembler (note how LLVM tries to return both y1 and y4 in the same eax register).

Will rustcc be able to detect and report that somehow?

Reply:
That seems to be a bug in LLVM. It should not allocate the *H registers for inline assembly.

Reply:
The emitted code is correct, though: after the assembly block, it first stores `al` somewhere else, only then `ah` in the entire `eax` register. Technically, since you only asked for the short registers, you shouldn't have a way to e.g. zero-extend one of them and clobber the other. Should this be a strict guarantee, that you don't get the `*h` registers?

Reply:
Yes, the way this RFC is currently written the `*h` registers are never chosen for operands (and I don't intend to change it).

The reason why I argue that LLVM's behavior is buggy is that Clang's inline asm is designed to emulate GCC's inline asm, and GCC never allocates `*h` for register operands.

I will submit a fix to LLVM as part of the `asm!` work.